File: Data Structures & Algorithms/linked-list-cycle-detection/submission-4.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} head
     * @return {boolean}
     */
    hasCycle(head) {
        //1. using set
        // if(!head) return false;
        // let set=new Set(), curr=head;
        // while(curr && curr.next){
        //     if(set.has(curr)) return true;
        //     set.add(curr);
        //     curr=curr.next;
        // }
        // return false;

        //using two pointer 
        if(!head) return false;
        let slow=head, fast=head;
        while(fast && fast.next){
            slow=slow.next;
            fast=fast.next.next;
            if(slow==fast) return true;
        }
        return false;
    }
}

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-4.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        // if(head==null) return head;
        // int length=0;
        // ListNode curr=head;
        // while(curr!=null){
        //     length++;
        //     curr=curr.next;
        // }
        // int index = length-n;
        // if(index == 0) return head.next;

        // curr=head;
        // for(int i=1;i<index;i++){
        //     curr=curr.next;
        // }
        // if(curr.next!=null){
        //     curr.next = curr.next.next;
        // }
        // else curr.next = null;
        // return head;

        //using slow & fast pointer
        if(head==null) return head;
        ListNode slow=head, fast=head;
        for(int i=0;i<n;i++){
            fast = fast.next;
        }

        if(fast == null) return head.next;
        while(fast!=null && fast.next!=null){
            slow= slow.next;
            fast= fast.next;
        }
        if(slow.next!=null){
            slow.next = slow.next.next;
        }
        else slow.next=null;
        return head;
    }
}
